File: src/problems_all/496_next_greater_element_i/496_next_greater_element_i_v1.go
/*
https://leetcode.com/problems/next-greater-element-i/


Runtime: 0 ms
Memory: 3.6 MB

https://leetcode.com/problems/next-greater-element-i/submissions/871877680/
*/

package main

import (
	"fmt"
	"math"
)

func main() {
	nums1 := []int{4, 1, 2}
	nums2 := []int{1, 3, 4, 2}
	result := nextGreaterElement(nums1, nums2)
	fmt.Println(result)
}

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	n2items := map[int]int{}
	for i, v := range nums2 {
		n2items[v] = i
	}

	result := []int{}
	for _, i := range nums1 {
		idx := n2items[i]
		nextBig := math.Inf(1)

		for j := idx; j < len(nums2); j++ {
			if nums2[j] > int(nextBig) && nums2[j] > i {
				nextBig = float64(nums2[j])
				break
			}
		}

		if nextBig == math.Inf(1) {
			result = append(result, -1)
		} else {
			result = append(result, int(nextBig))
		}
	}

	return result
}

File: src/problems_all/496_next_greater_element_i/496_next_greater_element_i_v1.py
"""
https://leetcode.com/problems/next-greater-element-i/


Runtime: 38 ms
Memory: 14.2 MB

https://leetcode.com/problems/next-greater-element-i/submissions/871869744/
"""


from typing import List


class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        n2items = {}
        for i, v in enumerate(nums2):
            n2items[v] = i

        result = []
        for i in nums1:
            idx = n2items[i]
            next_big = float('-inf')
            for j in range(idx, len(nums2)):
                if nums2[j] > next_big and nums2[j] > i:
                    next_big = nums2[j]
                    break

            if next_big == float('-inf'):
                result.append(-1)
            else:
                result.append(next_big)

        return result


def main():
    nums1 = [4, 1, 2]
    nums2 = [1, 3, 4, 2]
    result = Solution().nextGreaterElement(nums1, nums2)
    print(result)


if __name__ == '__main__':
    main()
